File: f90code/CHP11/volume_region.f90
!
! Numerical Mathematics and Computing, Third Edition
! Ward Cheney & David Kincaid
! Brooks/Cole Publ. Co.
! ISBN 0-534-20112-1
! Copyright (c) 1994.  All rights reserved.
! For educational use with the Cheney-Kincaid textbook.
! Absolutely no warranty implied or expressed.
!
! Section 11.2
!
! File: volume_region.f90
!
! Volume of a complicated region in three-space by Monte Carlo
      
program volume_region
      integer, parameter :: n = 5000, iprt = 1000
      real, dimension(n,3) :: r
      integer ::i,m
      real x, y, z, vol
      call random_number(r)
      do i = 1,n
         x = r(i,1); y = r(i,2); z = r(i,3)
         if((x*x+sin(y) <= z) .and. (x-z+exp(y) <= 1.0)) m = m + 1
         if(mod(i,iprt) == 0) then
            vol = real(m)/real(i)
            print *,i,vol
         end if
      end do
end program volume_region
  

File: c++code/CHP11/volume_region.c
/****************************************************************/
/* Module : volume_region.c					*/
/* Section: 11.2 				                */
/* Cheney-Kincaid, Numerical Mathematics and Computing, 4th ed, */
/* Brooks/Cole Publ. Co.                                        */
/* ISBN 0-534-20112-1                                           */
/* Copyright (c) 1999.  All rights reserved.                    */
/* For educational use with the Cheney-Kincaid textbook.        */
/* Absolutely no warranty implied or expressed.                 */   
/* 								*/
/* Programmer: Hidajaty Thajeb					*/
/* Description: Compute volume of a complicated region in       */
/*              3-space by Monte Carlo technique.		*/
/* Comment: The results might be slightly different from the    */
/*          book answers due to the different random values	*/
/*								*/
/****************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>


int main()
{
  const long MAXNUM = 2147483647;
  const int n = 5000,
            iprt = 1000;
  double r[n+1][3+1];
  int i, j, m;
  double vol, x, y, z;

  for (i = 1; i <= n; i++)
    for (j = 1; j <= 3; j++)   
      r[i][j] = (double)rand()/(double)MAXNUM;    
  m = 0;
  vol = 0;
  for (i = 1; i <= n; i++)
    {
      x = r[i][1];
      y = r[i][2];
      z = r[i][3];
      if ( ( x*x + sin(y) <= z) && (x - z + exp(y) <= 1))
	m++;
      if ( (i%iprt) == 0)
	{
	  vol = (double)m/(double)i;
	  printf("i = %d, vol = %.3f\n", i, vol);
	}
    }
  return 0;
}

